Add parser for more general intervals (slices)

NeilZaim opened the pull request:
Following issue #744, this PR adds a parser for having more general intervals for the inputs in `_int`.

**Proposed syntax**

This parser reads strings of the form `1:2:3, 4: :,5:6,  :  ,etc.`. A comma is used as a separator between different slices and whitespaces are trimmed and not taken into account.
A single slice can have 0, 1 or 2 colons.
* For 0 colon the given value is the period
* For 1 colon the given string is of type start:stop
* For 2 colons the given string is of type start:stop:period

Any value that is not given is set to default. Default is `0` for the start, `std::numeric_limits<int>::max()` for the stop and `1` for the period. For the 1 and 2 colon syntax, actually having the integers in the string is optional (this means that `::5`, `100 ::10` and `100 :` are all valid syntaxes).

**Syntax examples**

* `something_int = 50` -> do something at timesteps 0, 50, 100, 150, etc. (equivalent to `something_int = ::50`)
* `something_int = 300:600:100` -> do something at timesteps 300, 400, 500 and 600.
* `something_int = 300::50` -> do something at timesteps 300, 350, 400, 450, etc.
* `something_int = 105:108,205:208` -> do something at timesteps 105, 106, 107, 108, 205, 206, 207 and 208. (equivalent to `something_int = 105 : 108 : , 205 : 208 :`)
* `something_int = :` or  `something_int = ::` -> do something at every timestep.
* `something_int = 167:167,253:253,275:425:50` do something at timesteps 167, 253, 275, 325, 375 and 425.

You get the gist of it. This is essentially the python slicing syntax except that the stop is inclusive (`0:100` contains 100) and that no colon means that the given value is the period (this ensures backward compatibiilty).

**Implementation**

Everything was added in `WarpXUtil`. Let me know if you think it should have its own file.

First, a class `SliceParser` (feel free to propose other names throughout the PR) is defined, which corresponds to a single slice of the form `i:j:k`. This class has 3 int attributes corresponding to start, stop and period. It has a constructor taking a string as input and a method `contains` which takes an int `n` and returns `true` if the slice contains `n`. If the period is zero or negative, `contains` always returns `false` (as is the case currently: writing `something_int = -1` deactivates something).

Then, a class `MultiSliceParser` is also defined. It contains a vector of `SliceParser`. It has a constructor taking a string as input and a method `contains` which takes an int `n` and returns `true` if any of the slices contained in the `MultiSliceParser` contains `n`. I've also put a method `isActivated` which returns `true` if any of the slices contained in the `MultiSliceParser` has a strictly positive period. I figured it could be useful to replace the `if` statements that are currently of the form:
`if (something_int > 0) do_some_initialization_on_something()`.

I've also had to add a `split` function in the `WarpXUtilStr` namespace. It is similar to `amrex::Tokenize` but not exactly the same. The main difference is that, if `:` is used as a separator, `amrex::Tokenize` will split `":3::2"` into `["3","2"]` while this function will split `":3::2"` into `["","3","","2"]` (which is useful for differentiating `100`, `:100:`, `100:`, etc.). The `split` function can optionally trim the output split string using `amrex::trim()`.

In principle, only the constructor and methods of `MultiSliceParser` should be used in practice. If you want, I can take a bit of time to make this work for the existing `_int` input parameters in WarpX (actually I wanted to implement it for load balancing as an example in this PR but it turns out that the period `load_balancing_int` is currently used in the costs calcuations so we would first have to define what happens when we use multiple load balancing frequencies).

ax3l commented:
This is wonderful, thanks a lot!

Thank you for the detailed description, do you like to copy half of it directly for the user-docs inside `Docs/Source/` as well? :+1:

MaxThevenet commented on Source/Utils/WarpXUtil.H:
```diff
@@ -183,6 +183,111 @@ namespace WarpXUtilStr
+ */
+class SliceParser
+{
+    friend class MultiSliceParser;
```
Maybe we could avoid having `MultiSliceParser` as a friend, what do you think? For instance, if `contains` is public (see below) could we get rid of this? If we put these classes in different files, we might need a forward declaration otherwise.

MaxThevenet commented on Source/Utils/WarpXUtil.H:
```diff
@@ -183,6 +183,111 @@ namespace WarpXUtilStr
+    */
+    SliceParser (const std::string& instr);
+
+    /**
+    * \brief A method that returns true if the input integer is contained in the slice. (e.g. if
+    * the list is initialized with "300:500:100", this method returns true if and only if n is
+    * 300, 400 or 500). If the period is negative or 0, the function always returns false.
+    *
+    * @param[in] n the input integer
+    */
+    bool contains (int n);
```
```suggestion
public:
    /**
    * \brief A method that returns true if the input integer is contained in the slice. (e.g. if
    * the list is initialized with "300:500:100", this method returns true if and only if n is
    * 300, 400 or 500). If the period is negative or 0, the function always returns false.
    *
    * @param[in] n the input integer
    */
    bool contains (const int n) const;
```

MaxThevenet commented on Source/Utils/WarpXUtil.H:
```diff
@@ -183,6 +183,111 @@ namespace WarpXUtilStr
+ */
+class MultiSliceParser
+{
+    std::vector<SliceParser> m_slices;
+    std::string m_separator = ",";
+    bool m_activated = false;
```
```suggestion
private:
    std::vector<SliceParser> m_slices;
    std::string m_separator = ",";
    bool m_activated = false;
```
Also maybe we could put these at the end, after the public members?

MaxThevenet commented on Source/Utils/WarpXUtil.H:
```diff
@@ -183,6 +183,111 @@ namespace WarpXUtilStr
+    * @param[in] instr an input string of the form "x,y,z,...". This will call the constructor of
+    * SliceParser using x, y and z as input arguments.
+    */
+    public:
```
```suggestion
public:
```
This applies to all following members, and usually doesn't have this extra indentation.

MaxThevenet commented on Source/Utils/WarpXUtil.H:
```diff
@@ -183,6 +183,111 @@ namespace WarpXUtilStr
+    *
+    * @param[in] n the input integer
+    */
+    bool contains (int n);
```
```suggestion
    bool contains (const int n) const;
```

MaxThevenet commented on Source/Utils/WarpXUtil.H:
```diff
@@ -183,6 +183,111 @@ namespace WarpXUtilStr
+    * \brief A method that returns true if any of the slices contained by the MultiSliceParser
+    * has a strictly positive period.
+    */
+    bool isActivated ();
```
```suggestion
    bool isActivated () const;
```

MaxThevenet commented on Source/Utils/WarpXUtil.cpp:
```diff
@@ -246,4 +246,62 @@ namespace WarpXUtilStr
+    }
+}
+
+bool SliceParser::contains (int n)
```
```suggestion
bool SliceParser::contains (const int n) const
```

MaxThevenet commented on Source/Utils/WarpXUtil.cpp:
```diff
@@ -246,4 +246,62 @@ namespace WarpXUtilStr
+bool SliceParser::contains (int n)
+{
+    if (m_period <= 0) {return false;}
+    else{
+    return (n - m_start) % m_period == 0 && n >= m_start && n <= m_stop;}
```
```suggestion
    else{
        return (n - m_start) % m_period == 0 && n >= m_start && n <= m_stop;}
```

MaxThevenet commented on Source/Utils/WarpXUtil.cpp:
```diff
@@ -246,4 +246,62 @@ namespace WarpXUtilStr
+    }
+}
+
+bool MultiSliceParser::contains (int n)
```
```suggestion
bool MultiSliceParser::contains (const int n) const
```

MaxThevenet commented on Source/Utils/WarpXUtil.cpp:
```diff
@@ -246,4 +246,62 @@ namespace WarpXUtilStr
+    return false;
 }
+
+bool MultiSliceParser::isActivated () {return m_activated;}
```
```suggestion
bool MultiSliceParser::isActivated () const {return m_activated;}
```

NeilZaim commented:
Thanks a lot for the feedback! 

Yes I wrote the two first sections of the PR description with the idea of adding them to the doc later on. However I thought it would make more sense to wait until we actually use the parser before adding this description to the doc, what do you think? I guess the last section can also go in the developers doc, although it's not as useful.

Otherwise I'm fine with moving the classes in separate files but having 4 extra files (2 `.H` and 2 `.cpp`) might bit a bit too much for this no? Since the two classes are relatively short and are related to each other I guess they could belong in the same file. What do you think?

(and I'll try to pay more attention to const-correctness in the future :) )

MaxThevenet commented:
Oh sure, feel free to add the last section in the developer's doc, thanks! I would say the documentation of the parser itself can already go in the doc. Then, when we start using it for some inputs, we will update the doc of these inputs and refer to the parser doc (as we do for the math parser).

>having 4 extra files (2 .H and 2 .cpp) might bit a bit too much for this no?

Both are fine, and in fact both approaches are used in the code (and I am sure we could have a great debate on this topic :) ). In this case, we could consider that the main class is `MultiSliceParser` (and maybe find a more straightforward name), so we could create only 1 `.H/.cpp` pair and write both classes in there, considering that `SliceParser` is just an auxiliary class. Since we don't expect to use `SliceParser` directly anywhere in the code, that sounds good too.

>(and I'll try to pay more attention to const-correctness in the future :) )

No problem, the PR was great so I was just trying to find something to say :).

NeilZaim commented:
So I've moved the classes to a separate file (named `IntervalsParser.H/cpp` ; once again name is open to suggestions) and I've added a section at the bottom of `parameters.srt` with the two first paragraphs in this PR description.

ax3l commented on Docs/source/running_cpp/parameters.rst:
```diff
@@ -1529,3 +1529,48 @@ WarpX supports checkpoints/restart via AMReX.
     Name of the checkpoint file to restart from. Returns an error if the folder does not exist
     or if it is not properly formatted.
+
+Intervals parser
+--------------------------------------
```
`.rst` formatting detail:
```suggestion
Intervals parser
----------------
```

MaxThevenet commented on Source/Utils/IntervalsParser.cpp:
```diff
@@ -0,0 +1,61 @@
+    return false;
+}
+
+bool IntervalsParser::isActivated () const {return m_activated;}
```
```suggestion
bool IntervalsParser::isActivated () const {return m_activated;}

```

ax3l commented on Docs/source/running_cpp/parameters.rst:
```diff
@@ -1529,3 +1529,48 @@ WarpX supports checkpoints/restart via AMReX.
+Intervals parser
+--------------------------------------
+
+WarpX provides an interval parser that reads expressions in the input file of the form
+``1:2:3, 4: :,5:6,  :  ,etc.``.
+A comma is used as a separator between different slices and whitespaces are trimmed and not taken
+into account.
+A single slice can have 0, 1 or 2 colons.
```
Let's keep it simple, especially with white spaces, for the first examples. Introduce one term at a time.
```suggestion
WarpX can parse interval expressions of the form ``start:stop:period``, e.g.
``1:2:3, 4::, 5:6, :, ::10``.
A comma is used as a separator between groups of intervals, which we call slices.
White spaces are ignored.
A single slice can have 0, 1 or 2 colons, just as `numpy slices <https://numpy.org/doc/stable/reference/generated/numpy.s_.html>`_, but with inclusive upper bound for ``stop``.
```

NeilZaim commented:
@MaxThevenet I agree let's implement this for load balancing (this will also be a good way to test the classes). Either in this PR or in a separate one is fine for me really.

As I've said I'm not really sure about how to replace the following line in the cost calculations in `WarpXEvolve.cpp` so you'll have to help me on this :D (or maybe @mrowan137 knows?) :

`(*cost)[i] *= (1. - 2./load_balance_int);`

Is the relevant parameter the number of timesteps since the last load_balance or until the next load_balance (in which case we can add a dedicated method in `IntervalsParser` for that)? Or is it something else?

@ax3l Thanks for the review! I agree with your suggestions and I'll commit them.

mrowan137 commented:
> 
> `(*cost)[i] *= (1. - 2./load_balance_int);`
> 
> Is the relevant parameter the number of timesteps since the last load_balance or until the next load_balance (in which case we can add a dedicated method in `IntervalsParser` for that)? Or is it something else?

@NeilZaim it's the number of timesteps until the next load balance, in general.

MaxThevenet commented:
Currently we only specify a period (called `load_balance_int` for the load balancer, but it has different names for other routines). Then we use it in the code, either to determine when to do the operation, for instance
```c++
if (step > 0 && (step+1) % load_balance_int == 0){
```
or to compute some stuff
```c++
(*cost)[i] *= (1. - 2./load_balance_int);
```
For the first one, when we're using the parser, we could replace the if statement with
```c++
if (interval_parser.isRequested(step)){
```
where `interval_parser` would be an instance of `IntervalParser`, and `isRequested` would return `true` if `step` is a number requested by the user, `false` otherwise.

For the second one, we could also have additional function `IntervalParser::nextRequested(current_step)` (respectively `IntervalParser::previousRequested(current_step)`) that would return the next step > current_step (respectively previous step < current_step) for which `isRequested(step) = true`. The second line would then like like
```c++
(*cost)[i] *= 1 - 2./(step - interval_parser.previousRequested(step));
```
These functions would probably prove handy in some other situations, and hopefully that's all we'd need.

NeilZaim commented:
Thanks for your replies. If that's ok for you I will add this in this PR.

Just to clarify, if I understood correctly Michael's reply your second example should be more like:

`(*cost)[i] *= 1 - 2./(interval_parser.nextRequested(step) - interval_parser.previousRequested(step));`

With one of the method being inclusive and the other exclusive. Am I correct?

And in the cases where we are before the first load balance/after the last load balance, does having `previousRequested`/`nextRequested` return `0`/`max_step` seem reasonable to you? (at least for this specific case)

MaxThevenet commented:
Sure, it's fine to do it in this PR.
As for Michael's example, I expected this piece of code to be executed only at LB steps, not for all steps. If so, I think the solution I proposed is correct. Otherwise, yours should be used. I am not too sure whether they should have a different behavior (inclusive/exclusive) though.

And yes, the default behavior you mention sounds good.

mrowan137 commented:
@NeilZaim @MaxThevenet I think we should apply the cost weighting at each step, if we want to keep the calculation of the moving average the same as it is now.  practically, I think either what Max suggested, or at each step doing `(*cost)[i] *= 1 - 2./(interval_parser.nextRequested(step) - interval_parser.previousRequested(step));` will work (i.e., give a similar load-balanced distribution mapping), but if we go with the former we will want to stop clearing the costs at each load balance interval (so this becomes a question of whether we want costs to have memory since the beginning of the simulation or just since the previous load balance).

NeilZaim commented:
Hi! I've commited an implementation of the parser for load balancing.

For the last point in the discussion, since this PR is about the parser I prefer not to change anything that concerns load balancing here. Currently the operation on the costs is done at every timestep so I've left it this way and, as discussed above, replaced `load_balance_int` by something that calculates `next(timestep) - previous(timestep)`.

One of the two methods `next`/`previous` needs to be inclusive and the other inclusive, otherwise this fails at timesteps where load balancing is done. (e.g. if `load_balance_int = 100`, at timestep `200` we don't want to obtain `300-100 = 200`)

In practice I've chosen `nextExclusive(timestep) - previousInclusive(timestep)` because the operation on the cost is done after load balancing (if performed at the considered timestep) so I figured it made more sense to always replace `load_balance_int` by the number of timesteps between the previous load balance and the next one.

To do this I've added a bunch of methods in the `SliceParser` and `IntervalsParser` classes but in this line of code, `load_balance_int` is just replaced by a call to a method called `localPeriod` that does the operation `nextExclusive(timestep) - previousInclusive(timestep)`. (if we ever need another method that does `nextInclusive(timestep) - previousExclusive(timestep)` this can of course be easily added as well)

I've made a few tests with different random values for `load_balance_int` and it all seems to work properly but it's definitely possible that there are corner cases that aren't covered yet.

MaxThevenet commented on Source/Utils/IntervalsParser.H:
```diff
@@ -0,0 +1,146 @@
+    int getStop () const;
+
+private:
+    int m_start = 0;
+    int m_stop = std::numeric_limits<int>::max();
```
Maybe I miss something... is this modified at any point? If not, could we use `0` and `std::numeric_limits<int>::max()` instead?

NeilZaim replied:
Yes (if I understood your question correctly), `m_start` and `m_stop` are modified in the constructor of this class and correspond to the 2 first integers in an input string of the form `500:1000:100` (when we do something like `amrex::is_it<int>(insplit[0], m_start)` this assigns the value in `insplit[0]` to `m_start`; this is not very visible, in particular since it's inside an assert).
